# 01 Import Libraries


In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02 Importing Data

In [2]:
path = r'C:\Users\bulle\Documents\Python-Instancart Basket Analysis'
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [3]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [4]:
df_test = pd.DataFrame()

In [5]:
df_test['mix'] = ['a','b',1, True]

In [6]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [23]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [7]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [8]:
df_nan=df_prods[df_prods['product_name'].isnull() == True]

In [9]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [10]:
df_prods.shape

(49693, 5)

In [11]:
df_prods_clean =df_prods[df_prods['product_name'].isnull() == False]

In [12]:
df_prods_clean.shape

(49677, 5)

In [13]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [14]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [16]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [15]:
df_prods_clean.shape

(49677, 5)

In [18]:
df_prods_clean_no_dups.shape

(49672, 5)

In [19]:
#Export df_prods_clean_no_dups as "products_checked.csv"
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products.checked.csv'))

# Task 4.5

In [20]:
#Q1 done with consistency checks with "df_prods" (see above)
#Q2 check df_prods with df.describe()
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [22]:
#Maximum value in "prices" column showed as '99999". 
#After sorting, it showed two items prices are incorrect.
df_prods.sort_values(by=["prices"], ascending=False)

product_id                      product_name  aisle_id  department_id  \
33666       33664             2 % Reduced Fat  Milk        84             16   
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
25580       25579     Naturally Smoked Trout Fillet        15             12   
19392       19391         Turkey Breast Tenderloins        49             12   
40490       40486                   Chicken Tenders        49             12   
...           ...                               ...       ...            ...   
17058       17058                 Sausage Scrambles        52              1   
40191       40187                       Toffee Bars        37              1   
6934         6934            Chubby Hubby Ice Cream        37              1   
20105       20104          Organic Brown Rice Flour        17             13   
1963         1963             Organic Pregnancy Tea        94              7   

        prices  
33666  99999.0  
21554  14900.0  
25580     25.0  
19392     25.0  
40490     25.0  
...        ...  
17058      1.0  
40191      1.0  
6934       1.0  
20105      1.0  
1963       1.0  

[49693 rows x 5 columns]

In [24]:
#Q3/Q4, there are no mixed-type columns in df_ords:
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [25]:
#5, check missing value in df_ords:
#6 Only column has missing value is "days_since_prior_order"
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

Q5, there are 206209 rows of missing data in the "days_since_prior_order"

In [26]:
# create a dataframe for the missing values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [27]:
df_ords_nan

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

Q6, seems like all the missing value are customers' first purchase.

In [32]:
df_ords_cust_list= df_ords_nan

In [28]:
#7 check for duplicate in df_ords
df_ords_dups = df_ords[df_ords.duplicated()]

In [29]:
#7, there are no duplicates in df_ords
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

In [33]:
df_ords_clean_no_dups = df_ords

In [30]:
#Export df_prods_clean_no_dups as "products_checked.csv"
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products.checked.csv'))

In [34]:
#Export df_ords_clean_no_dups as "ords_checked.csv"
df_ords_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders.checked.csv'))